### PR TITLE
Use `rcl_subscription_set_on_new_message_callback`

### PR DESCRIPTION
**Pull request by pojiro**
This is a just idea to change subscription's timeout dynamically.

### 目的

メッセージが飛んでくる時間的な頻度に応じ、動的に `timeout [us]` を変えることで応答を良くすることを目的とする。

### 効果の検証

1000us 固定でも、 message が届いていればすぐ返るのだから、本修正は不要に考えられるが timeout を変更することで応答がよくなるという傾向が見られている。

1000us 固定の場合に複数の subscription GenServer がある場合に遅くなる傾向があるので、各 Erlang プロセスへの ERTS のスケジューラーのリソース配分が影響しているのかもしれない。が、憶測にすぎない。この辺を突き止めて改善できるとよい。

### CODE REVIEW

**Comment by FelixPe** [comment]
Hello @pojiro, I wanted to take a moment to express my sincere appreciation for addressing the topic at hand. It brings me great pleasure to see your work on this. Would the newly introduced [callback](https://docs.ros.org/en/humble/p/rcl/generated/function_subscription_8h_1aa63a1641e1138b83f882a9322a134adb.html#exhale-function-subscription-8h-1aa63a1641e1138b83f882a9322a134adb) functions help simplify this? From the callback it would be one option to [send](https://www.erlang.org/doc/man/erl_nif#enif_send) a message to the subscription process. A major disadvantage woult be, that it is not compatible to foxy anymore. What is your opinion on this approach?

**Comment by pojiro** [comment]
@FelixPe 

Thank you for letting me know about the existence of that function, `rcl_subscription_set_on_new_message_callback`.
I implemented, included the function to this PR branch, from humble rclex can use the function. I left foxy rclex as it was.

@takasehideki 

FelixPe-san taught me the function which doesn't block. From humble we can use the functions. This makes rclex very fast when there are a lot of subscription processes.

File: src/rcl_subscription.c [hunk 1]
```diff
@@ -99,3 +103,57 @@ ERL_NIF_TERM nif_rcl_take(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[]) {
...
+    return enif_make_badarg(env);
+  if (!rcl_subscription_is_valid(subscription_p)) return raise(env, __FILE__, __LINE__);
+
+  ErlNifPid *pid_p = (ErlNifPid *)enif_alloc_resource(rt_callback_resource, sizeof(ErlNifPid));
+  if (enif_self(env, pid_p) == NULL) return raise(env, __FILE__, __LINE__);
+  enif_keep_resource(pid_p);
+
+  rcl_ret_t rc;
+  rc = rcl_subscription_set_on_new_message_callback(subscription_p, new_message_callback,
+                                                    (const void *)pid_p);
```
Comment (pojiro) [inline src/rcl_subscription.c:132]: subscription のメッセージ受信時コールバックとして `new_message_callback` を登録する
このコールバックは RMW のスレッドで実行される

ref. https://docs.ros.org/en/humble/p/rmw/generated/function_rmw_8h_1aea20ab8bae7e3085de0cbe06564b4e93.html#

File: src/rcl_subscription.c [hunk 2]
```diff
@@ -99,3 +103,57 @@ ERL_NIF_TERM nif_rcl_take(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[]) {
...
+
+#ifndef ROS_DISTRO_foxy
+static void new_message_callback(const void *user_data, size_t number_of_events) {
+  ErlNifPid *pid_p = (ErlNifPid *)user_data;
+
+  ErlNifEnv *env = enif_alloc_env();
+  enif_send(env, pid_p, env,
+            enif_make_tuple(env, 2, new_message, enif_make_int(env, number_of_events)));
+  enif_free_env(env);
+}
```
Comment (pojiro) [inline src/rcl_subscription.c:115]: ROS 2メッセージの到着を subscription プロセスに `{:new_message, N}` の Erlang メッセージ として投げる。

File: src/rcl_subscription.c [hunk 3]
```diff
@@ -99,3 +103,57 @@ ERL_NIF_TERM nif_rcl_take(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[]) {
...
+  if (!enif_get_resource(env, argv[0], rt_rcl_subscription_t, (void **)&subscription_p))
+    return enif_make_badarg(env);
+  if (!rcl_subscription_is_valid(subscription_p)) return raise(env, __FILE__, __LINE__);
+
+  ErlNifPid *pid_p = NULL;
+  if (!enif_get_resource(env, argv[1], rt_callback_resource, (void **)&pid_p))
+    return enif_make_badarg(env);
+
+  rcl_ret_t rc;
+  rc = rcl_subscription_set_on_new_message_callback(subscription_p, NULL, NULL);
```
Comment (pojiro) [inline src/rcl_subscription.c:152]: `NULL` を登録することでコールバックをクリアできる。

> The callback to be called when new messages arrive, can be NULL to clear the registered callback

ref. https://docs.ros.org/en/humble/p/rmw/generated/function_rmw_8h_1aea20ab8bae7e3085de0cbe06564b4e93.html#

File: lib/rclex/subscription.ex [hunk 4]
```diff
@@ -38,59 +38,111 @@ defmodule Rclex.Subscription do
...
-          case Nif.rcl_take!(state.subscription, message) do
-            :ok ->
-              message_struct = apply(state.message_type, :get!, [message])
-
-              {:ok, _pid} =
-                Task.Supervisor.start_child(
-                  {:via, PartitionSupervisor, {Rclex.TaskSupervisors, self()}},
-                  fn -> state.callback.(message_struct) end
-                )
+  case System.fetch_env!("ROS_DISTRO") do
```
Comment (pojiro) [inline lib/rclex/subscription.ex:56]: `ROS_DISTRO` に応じて、定義する関数を変える。

**Review by pojiro — COMMENTED** [review]
@takasehideki 本PRの修正内容に関する説明を以下に付けました。

**Comment by takasehideki** [comment]
@FelixPe Thank you so much for sharing new knowledge! I did not know about this new function either:D

@pojiro I'm so happy if this PR could be merged and released as the new version (v0.10.1). Please let me know if you are ready (I think this is still WiP for you).

One thing that bothers us is that the code has diverged because we still have to support Foxy. I don't think now is the time, but I plan to drop Foxy support in the next release that moves up a minor number (v0.11.x).

**Comment by pojiro** [comment]
@takasehideki

I make this PR is ready to review and merge.
I would like to include some fixes found through measurements when I'm doing. So please wait for the release of 0.10.x.
I make an other PR which include the fixes.

**Comment by takasehideki** [comment]
@pojiro thank you so much for your effort!
OK, I merge this, but don't create the release.

**Pull request merged**